### DISCUSSION TITLE
Alt Builder for Architecting Forms via Schema

### DISCUSSION THREAD

**Discussion by zgover** [data-driven-forms/react-forms#1121] [Ideas]
It would be great to be able to have a drop-in component(s) for dynamic visual form building with greater support of  composition in a production ready env. Aside from than the existing builder which ultimately just demonstrates a playground to evaluate the library.

Ideas for purpose of developer benefits:
* limitations of form and fields 
* form renderer similar to the fields
* component mapper with default configurations and optionally permit overriding of independent variations of the derived props in field schema entry
* overrides component wrappers and which component 
* schema field entry minimizing total bytes of schema size schema data to avoid redundancy of commonly shared props of all fields or any field component
* permit optional component specification in field component propertyschema  compdata 
* specificity of form field values in→field values out→generated object result of cherry picks when form is holds dependencies of field values for dynamic, but not to forms objective output result. 

This could then ultimately be used regenerate the existing builder which could be renamed to Playground Demo and just be shown in docs of core library secondly shown in Builder docs for example of rendered final use case.

Any shared interests?

**Comment by Hyperkid123** [comment]
@zgover Thanks for the great ideas. The form builder is something we have wanted to complete for a very long time. So there is definitely a shared interest.

Almost every bullet you have posted is something I wanted to add to the current form build and ofc make it production-ready.

There is one big problem on our (myself and @rvsia) side. We just don't have the time to work on it. The current state of the builder is more than enough for our internal projects and there is so much other stuff to work on it would be very difficult to focus on the builder right now.

So yeah that is our status. I would love to spice up the builder, but just don't have the time. it is not a trivial feature.

 But if you @zgover have some spare resources I would be more than happy to join our brains.

**Comment by zgover** [comment]
@Hyperkid123 Yep, let's first setup an ability to whiteboard and identify a fixed list of imperative features and their design of logical flow to understand breaking changes to core library. I am incorporating a similar feature shortly into [Aglyn](https://aglyn.com) here soon, so it may provide me with further suggestions to logic.

**Reply by Hyperkid123** [reply]
I've created a GH project in the form builder repository: https://github.com/data-driven-forms/form-builder/projects/1. I've sent you an invite to the org so you should be able to add stuff to the board. Let me know if it does not work.

**Comment by rvsia** [comment]
I just want to add an important note here, we definitely wanted to introduce more features to the form builder, but there is no solid drag and drop library supporting **nested constructions**. And this was a blocker for us, because that's definitely one of the most important and basic features we need. And implementing a custom solution is very very time consuming. There is actually a work-in-progress branch with proof-of-concept of nesting supported, but we haven't touched it for months now.

**Reply by zgover** [reply]
> but there is no solid drag and drop library supporting nested constructions

I disagree, have to looked into [@dnd-kit/core](https://docs.dndkit.com/)

**Reply by rvsia** [reply]
Yeah, this one we are using in the [wip branch](https://github.com/data-driven-forms/form-builder/tree/dnd-kit) - we created a POC that there can be nested objects, which is fine, but there is still a long way to get it to a functional state (animations, keyboard support, updated collisions, etc.) And still, it wasn't so simple to customize it and the documentation wasn't great at that time.

**Comment by zgover** [comment]


**Reply by Hyperkid123** [reply]
I will try to read through it today.

**Reply by Hyperkid123** [reply]
@zgover So I was looking a bit at the library. 

We basically have this already implemented.
- The schema builder is a reducer function. It's just a different syntax to the effector one.
- we also have a customizable layout. You have 3 three different sections. One for the preview, one for inserting a new field, and one for the field properties editor.
- There is a "save schema" option. I think it is the same is the "save form" option in the snippet. Granted the formats are different, but transforming the DDF schema (which will be the output format) is a simple task.
- the CSS section is an equivalent if the component as I understand or maybe just to make the demo look a bit better

Basically, the only thing right now in the dnd kit version is to finish the drag and drop logic, mainly the collision detection. That is the current blocker.

Or should I understand it as you would also like to see a simpler version of the builder? Meaning that instead of DnD, just some static create field form which would just append a field to the end of the current list? And then maybe have some order features.

**Reply by zgover** [reply]
I read through it again and actually realized was mistaken and it is essentially the same thing. I must have interpreted something differently upon initially skimming. Removed to avoid confusion.
